Turn off log depth for large globe tiles

OmarShehata opened the pull request:
This fixes https://github.com/AnalyticalGraphicsInc/cesium/issues/7189 in the same way we fixed it for ClassificationPrimitives in https://github.com/AnalyticalGraphicsInc/cesium/pull/6465.

In master:

![](https://user-images.githubusercontent.com/1711126/47519804-66516000-d85c-11e8-8307-b36d8215521a.gif)

In this branch:

![](https://user-images.githubusercontent.com/1711126/47519816-70735e80-d85c-11e8-8a63-4415becac34f.gif)

_These gifs are copied from that issue, but this is indeed what I see in this branch._

This is just from opening the [3D Tiles BIM](https://cesiumjs.org/Cesium/Build/Apps/Sandcastle/?src=3D%20Tiles%20BIM.html) Sandcastle with no modification.

I noticed there was already a condition for ~disabling log depth~ pushing depth writing to the fragment shader in `GlobeSurfaceShaderSet`. I added one more check, to use the `isReadyForAtmosphere` boolean, which is computed in Scene.js and is true only when enough globe tiles have loaded to cover the screen and the atmosphere renders.

Code is NOT final. I threw the `isReadyForAtmosphere` in FrameState so I can pass it to GlobeSurfaceShaderSet, but I wanted to get feedback on whether this was a good approach first. @bagnell can you take a look?

Todo:

* [x] Update changes
* [x] Cleanup code, we probably don't want this `isReadyForAtmosphere` on frameState.
* [x] Test this with the EXT_frag_depth extension disabled.

cesium-concierge commented:
Thanks for the pull request @OmarShehata!
* :heavy_check_mark: Signed CLA found.
* :grey_question: [CHANGES.md](https://github.com/AnalyticalGraphicsInc/cesium/blob/log-depth/CHANGES.md) was not updated.
    * If this change updates the public API in any way, please add a bullet point to `CHANGES.md`.
* :grey_question: Unit tests were not updated.
    * Make sure you've [updated tests](https://github.com/AnalyticalGraphicsInc/cesium/tree/master/Documentation/Contributors/TestingGuide) to reflect your changes, added tests for any new code, and ran the code coverage tool.

Reviewers, don't forget to make sure that:

- [ ] Cesium Viewer works.
- [ ] Works in 2D/CV.
- [ ] Works (or fails gracefully) in IE11.

bagnell commented on Source/Scene/Scene.js:
```diff
@@ -2858,6 +2858,7 @@ define([
             if (defined(skyAtmosphere) && defined(globe)) {
                 skyAtmosphere.setDynamicAtmosphereColor(globe.enableLighting);
                 environmentState.isReadyForAtmosphere = environmentState.isReadyForAtmosphere || globe._surface._tilesToRender.length > 0;
+                frameState.isReadyForAtmosphere = environmentState.frameState;
```
`environment.frameState` is always `undefined` so log depth is always disabled.

bagnell commented:
I tested this and it does work, but log depth is never turned back on. I changed the line you added in Scene to below, but the issue is still there.

`frameState.isReadyForAtmosphere = environmentState.isReadyForAtmosphere;`

The better solution would be to modify this line to include some heuristic based on the camera distance to the surface and number of tiles rendered.

`environmentState.isReadyForAtmosphere = environmentState.isReadyForAtmosphere || globe._surface._tilesToRender.length > 0;`

For example, if the camera is viewing the entire globe, the the number of tiles should be 2 for the two root tiles. If the camera is close to the surface, like in the screen shot above, then wait for more tiles to be ready, like 10(?).

OmarShehata commented:
@bagnell ok, I've fixed the issue where log depth was never getting turned back on, and also simplified and cleaned up the code. 

It is not enough to simply not render the atmosphere, because there'll still be a black hole in the Earth, and it becomes jarring to the user when the sky goes from black to blue after a second instead of a frame. 

What it does now is it will wait until enough tiles load, and turn on log depth after. The magic number "40" was found by empirically starting with 1 and binary searching until I found the minimum that produced no gaps regardless of how close or far you are to the surface. I considered checking for the camera height, but I couldn't find any artifacts arising due to this. 

[Here's a Sandcastle](http://localhost:8080/Apps/Sandcastle/index.html#c=rVXbjtNADP2VqC+0Upl6PDcPlBWwAgkJCSRuD5SH0AzbEWlSTdJWC9p/x2nSpQssAkSeMh7bxz4+cXZ5ynYx7EPKHmRV2GfnoYnbtXh7sI3vLA/H87pq81iFdGeafV1UGT9tSIlNL1O9i0VI946ByxTyNryrU1m87l3Gk0V1Nbm/qBbVjtGGwGfV08RZX3x6WZeXF3XF8EPmIjRtrPI2svHeaU3neWr5La/U+C4qYxC1MNoQaQI/ze4qpyWhEV4bh0YDTTO+VZZQWL4zDpyeTHuQOsXAPQ0gA/IBPaaw/D02CG3AG4ZBb7Umqyyjg7CalDSotSICdL1NKm2k5WIs16b8Eb97tpvfYSjkyqUBo4wEMEYf8hmUDh0BgVLSazaCcOQVolNKOdSW6Ihxxbxf0z7QfE37MJy/p91LZD6d9eSlk67n3SkHAkCiRG/MgXcw5IRHy06g/gPtzLpDr6xhOGSKQZlD8wxjJTppeQyaqLNpzZQ4NqG03hur/5x1HpElT2S04/aM66eoHCnnLE9SIpDCDoQ8DxVAeemBR/VL0gPT239dgtuObQyNyItiPDS+GaR/g4lViBerlm1oAU7qXsWQ8rRcXd507/M0seOu4av3t0+OpCJBPCsNSLIfnAaHLGdvtOTTYXBSGhLGSc39K2dOufvpuR3MoZJCe+QUYEH1aKxkxXIlawElDTpxgMJ3VFu06l/hOlkKHgSC0erYG2ovAFkJpE3fG3LrQrKLoxNV/nVrQJyXV4DB/ttnMC5BsfAQWIPyyKS3UvB6QG4Y/eTDd7SrE+Q178sU8zK73qHndVkn8fj5mydiH9vVo3Kzyln/ZvJdYsM+7bW1zNesDdGEtlvb41s2bBcym/065Jbt0KPsY1XUe3H9m+hf7o+mo3nTXpbhrC/rYVxv6tRm21SOhZi1Yb0pubVm9nG7/BxasWyaLt98dgyaF3GXxeLBYvTDX2YxypZl3jR882lblq/il7AYnc1n7H8jrKzzIlYXL3Yhlfll57KSZ897oxBiPuPjz1FtXZcf83SS8Rs) showing a polygon correctly occluded by terrain. You can swap out the view line to get a view with the polygon in front of terrain instead. I don't see any depth issues in this branch. Let me know if there are other scenarios to check.

bagnell commented:
> I considered checking for the camera height, but I couldn't find any artifacts arising due to this.

You can't just pick 40 tiles, since there won't always be 40 tiles rendered in a view. For example, go to the home view. There will only ever be 1-4 tiles rendered so the sky atmosphere from space will never render.

OmarShehata commented:
> You can't just pick 40 tiles

Good point. Although this won't change whether the sky atmosphere is rendered, it'll just change whether the surface tiles are using log depth. 

I tried doing it based on height, but the issue is you could be very close to the ground, and be looking straight down, and then there's be only 5-6 tiles that are rendered. Doing it based on distance from the camera to the surface along the view vector is also misleading because you could be looking at the horizon (or at the sky) and then the distance is very large, but the camera could be very close to the ground. 

What I just pushed is two things:

* I noticed the ClassificationPrimitive fix was changed from to `DISABLE_GL_POSITION_LOG_DEPTH` to `ENABLE_GL_POSITION_LOG_DEPTH_AT_HEIGHT` in  https://github.com/AnalyticalGraphicsInc/cesium/commit/b5ba89342588ec7d3e7e6858f3579af237d38a05#diff-1e12b809c9b98d1f389fdc6546af07c4, so I changed it to that as well here. (I'm assuming this will prevent the same kind of artifact here?)
* I am using a conservative estimate here of a maximum of 10 tiles. It should keep log depth on in most views except when you're very close to the ground for a few seconds. 

The caveats are:

* It is still possible that there will be views that don't have enough tiles to trigger log depth.
* It updates every frame, so if you change terrain providers it still works, which is good, but it also means you might move into a view that has less terrain tiles and triggers log depth off.

@bagnell what do you think? This is really is a temporary solution to the log depth with large geometries issue. I haven't seen any bad artifacts that come as a result of this yet.

cesium-concierge commented:
Thanks again for your contribution @OmarShehata! 

No one has commented on this pull request in 30 days. Maintainers, can you review, merge or close to keep things tidy?

 _I'm going to re-bump this in 30 days. If you'd like me to stop, just comment with `@cesium-concierge stop`. If you want me to start again, just delete the comment._

OmarShehata commented:
My suggestion to finishing this PR is to be more aggressive about the number of tiles required to trigger log depth, since I haven't been able to confirm any bad effects of erring on the side of turning off log depth on terrain temporarily here.

cesium-concierge commented:
Thanks again for your contribution @OmarShehata! 

No one has commented on this pull request in 30 days. Maintainers, can you review, merge or close to keep things tidy?

 _I'm going to re-bump this in 30 days. If you'd like me to stop, just comment with `@cesium-concierge stop`. If you want me to start again, just delete the comment._

cesium-concierge commented:
Thanks again for your contribution @OmarShehata! 

No one has commented on this pull request in 30 days. Maintainers, can you review, merge or close to keep things tidy?

 _I'm going to re-bump this in 30 days. If you'd like me to stop, just comment with `@cesium-concierge stop`. If you want me to start again, just delete the comment._

cesium-concierge commented:
Thanks again for your contribution @OmarShehata! 

No one has commented on this pull request in 30 days. Maintainers, can you review, merge or close to keep things tidy?

 _I'm going to re-bump this in 30 days. If you'd like me to stop, just comment with `@cesium-concierge stop`. If you want me to start again, just delete the comment._

hpinkos commented:
@OmarShehata what's the plan for this PR?

OmarShehata commented:
I think we should close but maybe keep the branch. Ideally we'd fix the underlying problem, but this stop-gap may be good if we can spend some more time testing it, coming up with a more robust heuristic based on camera distance to surface and angle, and make sure that this dynamically flipping logarithmic depth on/off doesn't cause any other artifacts.
